caffe2/operators/quantized/int8_add_op.h

```diff
@@ -1,219 +1,94 @@
-    auto X1_offset = -X1.zero_point;
-    int32_t Y_offset = this->template GetSingleArgument<int>("Y_zero_point", 0);
-    auto Y_scale = this->template GetSingleArgument<float>("Y_scale", 1);
-    const double twice_max_input_scale = 2 * std::max(X0.scale, X1.scale);
-    const double real_X0_multiplier = X0.scale / twice_max_input_scale;
-    const double real_X1_multiplier = X1.scale / twice_max_input_scale;
-    const double real_Y_multiplier =
-        twice_max_input_scale / ((1 << kAddLeftShift) * Y_scale);
-
-    Y->t.ResizeLike(X0.t);
-    Y->zero_point = Y_offset;
+
+    const int32_t Y_zero_point = this->template GetSingleArgument<int>("Y_zero_point", 0);
+    const float Y_scale = this->template GetSingleArgument<float>("Y_scale", 1);
+    Y->t.ResizeLike(A.t);
+    Y->zero_point = Y_zero_point;
     Y->scale = Y_scale;
 
-    int32_t X0_multiplier;
-    int X0_shift;
-    QuantizeMultiplierSmallerThanOne(
-        real_X0_multiplier, &X0_multiplier, &X0_shift);
-    int32_t X1_multiplier;
-    int X1_shift;
-    QuantizeMultiplierSmallerThanOne(
-        real_X1_multiplier, &X1_multiplier, &X1_shift);
-    int32_t Y_multiplier;
-    int Y_shift;
-    QuantizeMultiplierSmallerThanOne(
-        real_Y_multiplier, &Y_multiplier, &Y_shift);
-
-    Int8Add(
-        X0.t.template data<uint8_t>(),
-        X0.t.numel() / X0.t.size(X0.t.dim() - 1),
-        X0.t.size(X0.t.dim() - 1),
-        X0_offset,
-        X0_multiplier,
-        X0_shift,
-        X1.t.template data<uint8_t>(),
-        X1_offset,
-        X1_multiplier,
-        X1_shift,
-        Y_offset,
-        Y_multiplier,
-        Y_shift,
+    initQNNPACK();
+
+    pthreadpool_t threadpool =
+        reinterpret_cast<pthreadpool_t>(ws_->GetThreadPool());
+
+    if (this->qnnpackOperator_ == nullptr) {
+      const qnnp_status createStatus = qnnp_create_add_nc_q8(
+        1 /* channels */,
+        A.zero_point, A.scale,
+        B.zero_point, B.scale,
+        static_cast<uint8_t>(Y_zero_point), Y_scale,
+        activationLimits(Y_scale, Y_zero_point, Ac).first,
+        activationLimits(Y_scale, Y_zero_point, Ac).second,
+        &qnnpackOperator_);
+      CAFFE_ENFORCE(
+          createStatus == qnnp_status_success,
+          "failed to create QNNPACK add operator");
+      CAFFE_ENFORCE(this->qnnpackOperator_ != nullptr);
+    }
+
+    const qnnp_status setupStatus = qnnp_setup_add_nc_q8(
+        this->qnnpackOperator_,
+        A.t.numel() /* batch size */,
+        A.t.template data<uint8_t>(),
+        1 /* A stride */,
+        B.t.template data<uint8_t>(),
+        1 /* B stride */,
         Y->t.template mutable_data<uint8_t>(),
-        activationLimits(Y->scale, Y->zero_point, Ac).first,
-        activationLimits(Y->scale, Y->zero_point, Ac).second,
-        ws_->GetThreadPool());
+        1 /* Y stride */);
+    CAFFE_ENFORCE(
+        setupStatus == qnnp_status_success,
+        "failed to setup QNNPACK add operator");
+
+#ifdef FBCODE_CAFFE2
```

Comment:
Why is this branch needed?

Reply:
Because both `pthreadpool` and Caffe2 thread pool implement the same functions, and fbcode does not allow dependency on interface without dependency on implementation. QNNPACK depends on pthreadpool interface, therefore in fbcode it depends on pthreadpool implementation, and therefore we can't just pass Caffe2 thread pool pointer here.